hpat/hiframes/dataframe_pass.py

```diff
@@ -1357,6 +1372,31 @@ def _run_call_drop(self, assign, lhs, rhs):
         _init_df = _gen_init_df(out_typ.columns)
         return self._replace_func(_init_df, data, pre_nodes=nodes)
 
+    def _run_call_df_isna(self, assign, lhs, rhs):
+        df_var = rhs.args[0]
+        df_typ = self.typemap[df_var.name]
+
+        # impl: for each column, convert data to series, call S.isna(), get
+        # output data and create a new dataframe
+        n_cols = len(df_typ.columns)
+        data_args = tuple('data{}'.format(i) for i in range(n_cols))
+
+        func_text = "def _isna_impl({}):\n".format(", ".join(data_args))
+        for d in data_args:
+            func_text += "  {} = hpat.hiframes.api.init_series({})\n".format(d + '_S', d)
+            func_text += "  {} = hpat.hiframes.api.get_series_data({}.isna())\n".format(d + '_O', d + '_S')
+        func_text += "  return hpat.hiframes.pd_dataframe_ext.init_dataframe({}, None, {})\n".format(
+            ", ".join(d + '_O' for d in data_args),
+            ", ".join("'{}'".format(c) for c in df_typ.columns))
+
+        loc_vars = {}
+        exec(func_text, {}, loc_vars)
```

Comment:
Just to clarify the function generated here:
```
def _isna_impl(data0, data1, data2, data3):
  data0_S = hpat.hiframes.api.init_series(data0)
  data0_O = hpat.hiframes.api.get_series_data(data0_S.isna())
  data1_S = hpat.hiframes.api.init_series(data1)
  data1_O = hpat.hiframes.api.get_series_data(data1_S.isna())
  data2_S = hpat.hiframes.api.init_series(data2)
  data2_O = hpat.hiframes.api.get_series_data(data2_S.isna())
  data3_S = hpat.hiframes.api.init_series(data3)
  data3_O = hpat.hiframes.api.get_series_data(data3_S.isna())
  return hpat.hiframes.pd_dataframe_ext.init_dataframe(data0_O, data1_O, data2_O, data3_O, None, 'A', 'B', 'C', 'D')
```
It is not a request for changes but, usually, I put an example of generated data in such cases to comments.

Reply:
I looked how a multi-line comment just before exec looks, but it actually makes code look over complicated, so I decided that's it's better without it. Using simple "print(func_text)" is sufficient most of the times.